happidev_lyrics: Convert former apiseeds_lyrics to Happi.dev API

phw opened the pull request:
Apiseed is shutting down and gets integrated into [Happi.Dev](https://happi.dev). If you are logged in into the Apiseeds website it shows this message:

> We are migrating to Happi.Dev. Apiseeds will be shut down soon. If you want use our services please go to Happi.Dev.

Happi.dev provides a Music API with lyrics search. As I underdstand it the API seeds data might get integrated there, but the API itself is a bit different. Hence this PR builds upon the Apiseeds Lyrics plugin, but adopts the code to match the new API. In particular this now requires two requests to fetch the lyrics.

See:
- PICARD-2112
- https://community.metabrainz.org/t/apiseeds-lyrics-offline/518636

phw committed:
happidev_lyrics: Convert former apiseeds_lyrics to Happi.dev API

Sophist-UK commented on plugins/happidev_lyrics/happidev_lyrics.py:
```diff
@@ -0,0 +1,171 @@
+                PLUGIN_NAME, metadata['title'], lyrics_url))
+            path = urlparse(lyrics_url).path
+            album._requests += 1
+            self._request(album.tagger.webservice, path,
```
I recommend that lyrics requests are put at the front FIFO of the non-priority queue so that when you load an album (or several albums) the lyrics requests are not queued up behind the search requests.

phw replied:
Not sure I follow. Would the suggestion be to queue the lyrics requests with `important=True` (which would put them at the top of the queue)?

Sophist-UK replied:
Yes - `important=True` for lyric requests, `important=False` for search requests.

phw replied:
Addressed in b0ff1c874e16e1abb424693148032b4477bb0d62

Sophist-UK commented on plugins/happidev_lyrics/happidev_lyrics.py:
```diff
@@ -0,0 +1,171 @@
+            parse_response_type='json', priority=True, queryargs=queryargs)
+
+    def process_search_response(self, album, metadata, response, reply, error):
+        if error or (response and (not response.get('success', False) or not response.get('length', 0))):
```
To save half the requests if you Reload an album, I recommend caching the search responses (which are small - but not the lyrics responses which are large).

phw replied:
I'll merge this now as is as I want to have a working plugin deployed and my main concern is to replace apiseeds with a functionally equivalent version for happi.dev, otherwise we will receive continuous issue reports about apiseeds ceasing to work.

Someone can address caching the results later.

Sophist-UK replied:
Please feel free to plagiarise my caching code in albumartistwebsite plugin.

Sophist-UK commented on plugins/happidev_lyrics/happidev_lyrics.py:
```diff
@@ -0,0 +1,171 @@
+        artist = metadata['artist']
+        if not artist:
+            log.debug(
+                '{}: artist is missing, please provide a valid value'.format(PLUGIN_NAME))
```
"please provide a valid value"???

How about "artist is required to obtain happi.dev lyrics"?

phw replied:
Done in 0def93a7a9a4c7761d67cae19aac4074143979dd

Sophist-UK commented on plugins/happidev_lyrics/happidev_lyrics.py:
```diff
@@ -0,0 +1,171 @@
+        title = metadata['title']
+        if not title:
+            log.debug(
+                '{}: title is missing, please provide a valid value'.format(PLUGIN_NAME))
```
"please provide a valid value"???

How about "title is required to obtain happi.dev lyrics"?

Sophist-UK replied:
Or even better:
```
        artist = metadata['artist']
        title = metadata['title']
        if not (artist and title):
            log.debug(
                '{}: both artist and title are required to obtain happi.dev lyrics'.format(PLUGIN_NAME))
            return
```
as this avoid a situation where they get a message about needing artist, provide the artist and then get a second error about needing title.

phw replied:
Yes, let's change this.

But technically this shouldn't happen, if the data is not available at the time the plugin runs something has gone wrong. I think these comments had been there more for debugging than for anything else.

Sophist-UK replied:
hmmm. I think you are probably right that the plugin only runs after successful download of data from MB, and MB should never have a situation where a track doesn't have an Artist or Title. But Sods Law says that there will be an example at some point.

Sophist-UK commented on plugins/happidev_lyrics/happidev_lyrics.py:
```diff
@@ -0,0 +1,171 @@
+                     'In order to use Happi.dev you need to get a free API key at <a href="https://happi.dev">happi.dev</a>'
+PLUGIN_VERSION = '2.0'
+PLUGIN_API_VERSIONS = ['2.0', '2.1', '2.2', '2.3', '2.4', '2.5', '2.6']
+PLUGIN_LICENSE = 'MIT'
```
I don't have an issue with MIT license, however since this is a rewrite replacement not an edit, IMO we can change to a different license if we wish e.g. GPL.

phw replied:
I'd keep the license. I did basically only take the apiseeds lyrics and ran a search and replace over it, then modified the request / response handling as necessary.

Sophist-UK reviewed (changes requested):
Nice to see it only takes 2 API calls per track - I was worried that it might take 3 or more.

Sophist-UK commented on plugins/happidev_lyrics/happidev_lyrics.py:
```diff
@@ -0,0 +1,171 @@
+            lyrics = response['result']['lyrics']
+            metadata['lyrics'] = lyrics
+            log.debug('{}: lyrics loaded for track {}'.format(
+                PLUGIN_NAME, metadata['title']))
```
Since both artist and title are required, how about:
```
            log.debug('{}: lyrics loaded for track {} by {}, '.format(
                PLUGIN_NAME, metadata['title'], metadata['artist']))
```

phw replied:
Yes makes sense.

Technically artist might not be required, but I think I'd keep it as it is now. For the Apiseeds API it was clear, because both artist and title where a required parameter for the API endpoint.

But with the Happi.dev API it is different, as we use a search endpoint. The structured parameters only tell the endpoint to return tracks with lyrics, but otherwise there is just a freeform search field (the q parameter). There is no documentation on how this works, but I found searching for `"ARTIST" "TITLE"`  seemed to work reasonably well.

Technically that means just the title would also work. But looking for artist + title at least seems to be a bit more reliable.

Sophist-UK replied:
I suspect that there are many examples of completely different songs with the same title by different artists.

I think it is far less likely that an artist has two completely different songs with the same title.

Given that it is reasonable to expect Picard's MB based metadata to include both artist and title, I think it is absolutely the correct thing to do to require both to avoid false positives.

phw replied:
I improved the logging to include the artist in 0def93a7a9a4c7761d67cae19aac4074143979dd

phw committed:
happidev_lyrics: Run lyrcis request as important

Avoids requests to get actual lyrics being delayed after all other loading has finished

phw committed:
happidev_lyrics: Improved logging output
